augur/io/metadata.py

```diff
@@ -223,9 +230,33 @@ def read_metadata_with_sequences(metadata, fasta, seq_id_column, seq_field='sequ
 
         # Skip records that do not have a matching sequence
         if seq_id not in sequence_ids:
+            # Save unmatched metadata ids to report unmatched records if requested
+            unmatched_metadata_ids.add(seq_id)
             continue
 
         sequence_record = sequences[seq_id]
         record[seq_field] = str(sequence_record.seq).upper()
+        # Save processed sequence ids to be able to determine if sequences were unmatched
+        processed_sequence_ids.add(seq_id)
 
         yield record
+
+
+    unmatched_sequence_ids = sequence_ids - processed_sequence_ids
+    if unmatched_reporting is not DataErrorMethod.SILENT and (unmatched_metadata_ids or unmatched_sequence_ids):
+        unmatched_message = "The output may be incomplete because there were unmatched records."
+
+        if unmatched_metadata_ids:
+            unmatched_message += "\nThe following metadata records did not have a matching sequence:\n"
+            unmatched_message += "\n".join(sorted(unmatched_metadata_ids))
+
+        if unmatched_sequence_ids:
+            unmatched_message += "\nThe following sequence records did not have a matching metadata record:\n"
+            unmatched_message += "\n".join(sorted(unmatched_sequence_ids))
+
+        if unmatched_reporting is DataErrorMethod.WARN:
+            print_err(f"WARNING: {unmatched_message}")
+        elif unmatched_reporting is DataErrorMethod.ERROR:
+            raise AugurError(unmatched_message)
```

Comment:
Reorganized warnings and errors in 0b637d298c9b90f3f76f82f00af4e806bb63b960. 
Looks a little complicated because of the two different error methods, but I think still manageable?

Reply:
LGTM (though I didn't go thru the tests with a fine-toothed comb).